src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+
+            // Ask the language to determine which of the *matched* items it wants to select.
+            var matchingItems = itemsInList.Where(r => r.FilterResult.MatchedFilterText)
+                                           .Select(t => t.FilterResult.CompletionItem)
+                                           .AsImmutable();
+
+            var chosenItems = filterMethod(matchingItems, filterText);
+
+            var recentItems = _recentItems;
+
+            // Of the items the service returned, pick the one most recently committed
+            var bestItem = Session.GetBestCompletionItemBasedOnMRU(chosenItems, recentItems);
+            VSCompletionItem uniqueItem = null;
+            int selectedItemIndex = 0;
+
+            // Determine if we should consider this item 'unique' or not.  A unique item
+            // will be automatically committed if the user hits the 'invoke completion' 
+            // without bringing up the completion list.  An item is unique if it was the
+            // only item to match the text typed so far, and there was at least some text
+            // typed.  i.e.  if we have "Console.$$" we don't want to commit something
+            // like "WriteLine" since no filter text has actually been provided.  HOwever,
+            // if "Console.WriteL$$" is typed, then we do want "WriteLine" to be committed.
+            if (bestItem != null)
+            {
+                selectedItemIndex = itemsInList.IndexOf(i => Equals(i.FilterResult.CompletionItem, bestItem));
+                if (selectedItemIndex > -1 && bestItem != null && matchingItems.Length == 1 && filterText.Length > 0)
+                {
+                    uniqueItem = highlightedList[selectedItemIndex].CompletionItem;
+                }
+            }
+
+            // If we don't have a best completion item yet, then pick the first item from the list.
+            var bestOrFirstCompletionItem = bestItem ?? itemsInList.First().FilterResult.CompletionItem;
+
+            // Check that it is a filter symbol. We can be called for a non-filter symbol.
+            if (filterReason == CompletionFilterReason.Insertion &&
+                !Controller.IsPotentialFilterCharacter(typeChar) &&
+                !string.IsNullOrEmpty(filterText) &&
+                !Controller.IsFilterCharacter(bestOrFirstCompletionItem, typeChar, filterText))
+            {
+                return null;
+            }
+
+            var updateSelectionHint = Session.IsHardSelection(
+                        filterText, initialRoslynTriggerKind, bestOrFirstCompletionItem,
+                        _completionHelper, filterReason, recentItems, hasSuggestedItemOptions)
+                        ? AsyncCompletionData.UpdateSelectionHint.Selected
+                        : AsyncCompletionData.UpdateSelectionHint.SoftSelected;
+
+            if (selectedItemIndex == -1)
+            {
+                selectedItemIndex = 0;
+            }
```

Comment:
what is this for?

Reply:
selectedItemIndex is originally defined by itemsInList.IndexOf. If not found, it can be -1.
However, we should ask editor to select the first item (index = 0) in the case. Added a comment about this.




---
In reply to: [232510359](https://github.com/dotnet/roslyn/pull/29016#discussion_r232510359) [](ancestors = 232510359)